### PR TITLE
[api] Add JWT token decoding

### PR DESCRIPTION
**Pull request by Harshg999**
## What changes were proposed in this pull request?

- Decode JWT token
- Find or create user from payload data
- Update tests for existing user, new user, authentication scenarios

## TODOs

- Token persistance
- Secret key handling and decide algorithm

## How was this patch tested?

- Successfully implementing and running unit tests

### CODE REVIEW

File: desktop/core/src/desktop/auth/api_authentications_tests.py [hunk 1]
```diff
@@ -36,16 +40,46 @@ def setUp(self):
...
+        "Bearer eyJhbGciOiJSUzI1NiJ9.eyJhdWQiOlsid29ya2xvYWQtYXBwIiwicmFuZ2VyIl0sImV4cCI6MTYyNjI1Njg5MywiaWF0IjoxNjI2MjU2NTkzLCJpc3MiOiJDbG91ZGVyYTEiLCJqdGkiOiJpZDEiLCJzdWIiOiJ0ZXN0LXN1YmplY3QiLCJ1c2VyIjoidGVzdF91c2VyIn0.jvyVDxbWTAik0jbdUcIc9ZANNrJZUCWH-Pg7FloRhg0ZYAETd_AO3p5v_ppoMmVcPD2xBSrngA5J3_A_zPBvQ_hdDlpb0_-mCCJfGhC5tju4bI9EE9Akdn2FrrsqrvQQ8cPyGsIlvoIxrK1De4f74MmUaxfN7Hrrcue1PTY4u4IB9cWQqV9vIcX99Od5PUaNekLIee-I8gweqvfGEEsW7qWUM63nh59_TOB3LLq-YcEuaX1h_oiTATeCssjk_ee9RrJGLNyKmC0WJ4UrEWn8a_T3bwCy8CMe0zV5PSuuvPHy0FvnTo2il5SDjGimxKcbpgNiJdfblslu6i35DlfiWg"
+      }
+      
+      request = MagicMock(HEADERS=HEADERS)
+
+      # Existing user
+      jwt_decode.return_value = {
+        "username": "test"
+      }
+      assert_equal(1, User.objects.count())
```
Comment (romainr) [inline desktop/core/src/desktop/auth/api_authentications_tests.py:56]: User.objects.filter(username="test").exists()?
... etc

(will fail when running full test suite which creates other users)

File: desktop/core/src/desktop/auth/api_authentications_tests.py [hunk 2]
```diff
@@ -36,16 +40,46 @@ def setUp(self):
...
+
+      assert_equal(user, self.user)
+      assert_true(user.is_authenticated)
+      assert_equal(1, User.objects.count())
+
+      # New user
+      jwt_decode.return_value = {
+        "username": "test_new_user"
+      }
+      assert_equal(1, User.objects.count())
```
Comment (romainr) [inline desktop/core/src/desktop/auth/api_authentications_tests.py:68]: assert_false(...exists()) ?

File: desktop/core/src/desktop/auth/api_authentications_tests.py [hunk 3]
```diff
@@ -36,16 +40,46 @@ def setUp(self):
...
+      # Existing user
+      jwt_decode.return_value = {
+        "username": "test"
+      }
+      assert_equal(1, User.objects.count())
+
+      user, token = JwtAuthentication().authenticate(request=request)
+
+      assert_equal(user, self.user)
+      assert_true(user.is_authenticated)
```
Comment (romainr) [inline desktop/core/src/desktop/auth/api_authentications_tests.py:61]: To test that not a super user too

File: desktop/core/src/desktop/auth/api_authentications_tests.py [hunk 4]
```diff
@@ -36,16 +40,46 @@ def setUp(self):
...
+    with patch('desktop.auth.api_authentications.jwt.decode') as jwt_decode:
+
+      HEADERS = {
+        'HTTP_AUTHORIZATION':
+        "Bearer eyJhbGciOiJSUzI1NiJ9.eyJhdWQiOlsid29ya2xvYWQtYXBwIiwicmFuZ2VyIl0sImV4cCI6MTYyNjI1Njg5MywiaWF0IjoxNjI2MjU2NTkzLCJpc3MiOiJDbG91ZGVyYTEiLCJqdGkiOiJpZDEiLCJzdWIiOiJ0ZXN0LXN1YmplY3QiLCJ1c2VyIjoidGVzdF91c2VyIn0.jvyVDxbWTAik0jbdUcIc9ZANNrJZUCWH-Pg7FloRhg0ZYAETd_AO3p5v_ppoMmVcPD2xBSrngA5J3_A_zPBvQ_hdDlpb0_-mCCJfGhC5tju4bI9EE9Akdn2FrrsqrvQQ8cPyGsIlvoIxrK1De4f74MmUaxfN7Hrrcue1PTY4u4IB9cWQqV9vIcX99Od5PUaNekLIee-I8gweqvfGEEsW7qWUM63nh59_TOB3LLq-YcEuaX1h_oiTATeCssjk_ee9RrJGLNyKmC0WJ4UrEWn8a_T3bwCy8CMe0zV5PSuuvPHy0FvnTo2il5SDjGimxKcbpgNiJdfblslu6i35DlfiWg"
+      }
+      
+      request = MagicMock(HEADERS=HEADERS)
+
+      # Existing user
```
Comment (romainr) [inline desktop/core/src/desktop/auth/api_authentications_tests.py:52]: Move those to dedicated test methods?

Can move HEADERS into a class attribute to reuse it in the methods

**Review by romainr — COMMENTED** [review]
Super nice!

Just a few code organization polishing comments:

Title too generic and more noisy than helpful, e.g. should be more "[api] Add JWT token decoding"?

File: desktop/core/src/desktop/auth/api_authentications_tests.py (hunk 4 above)
Comment (Harshg999) [reply desktop/core/src/desktop/auth/api_authentications_tests.py:52]: Done

File: desktop/core/src/desktop/auth/api_authentications_tests.py (hunk 3 above)
Comment (Harshg999) [reply desktop/core/src/desktop/auth/api_authentications_tests.py:61]: Added

**Review by romainr — APPROVED** [review]
:+1:

**Pull request merged**